sbws/lib/relaylist.py

```diff
@@ -124,12 +125,26 @@ class RelayList:
     def __init__(self, args, conf, controller):
         self._controller = controller
         self.rng = random.SystemRandom()
+        self._refresh_lock = Lock()
         self._refresh()
 
+    def _need_refresh(self):
+        return time.time() >= self._last_refresh + self.REFRESH_INTERVAL
+
     @property
     def relays(self):
-        if time.time() >= self._last_refresh + self.REFRESH_INTERVAL:
-            self._refresh()
+        # See if we can get the list of relays without having to do a refresh,
+        # which is expensive and blocks other threads
+        if self._need_refresh():
+            # Whelp we couldn't just get the list of relays because the list is
+            # stale. Wait for the lock so we can refresh it.
+            with self._refresh_lock:
+                # Now we have the lock ... but wait! Maybe someone else already
+                # did the refreshing. So check if it still needs refreshing. If
+                # not, we can do nothing.
+                if self._need_refresh():
+                    self._refresh()
+        assert not self._need_refresh()
```

Comment:
There's a race condition here:
The first self._need_refresh() is executed at 299.99… seconds
The second self._need_refresh() is executed at 300.00… seconds

The probability of hitting this race condition is approximately (time needed to execute self._need_refresh()) / 300

You might want to remove the assert.

Reply:
(You could log a warning instead. Or you could assume that the probability is low enough.)

Reply:
Good point. Removed the assert.